larq_zoo/literature/meliusnet.py

```diff
@@ -0,0 +1,240 @@
+            padding="same",
+            use_bias=self.use_conv_bias,
+            input_quantizer=self.input_quantizer,
+            kernel_quantizer=self.kernel_quantizer,
+            kernel_constraint=self.kernel_constraint,
+            kernel_initializer=self.kernel_initializer,
+        )(x)
+
+    def group_conv(self, x, filters, kernel, groups):
+        assert filters % groups == 0
+        assert x.shape.as_list()[-1] % groups == 0
+
+        x_split = tf.split(x, groups, axis=-1)
+
+        y_split = []
+        for i in range(groups):
+            y_split.append(
+                tf.keras.layers.Conv2D(
+                    filters // groups,
+                    kernel,
+                    padding="same",
+                    use_bias=self.use_conv_bias,
+                    kernel_initializer=self.kernel_initializer,
+                )(x_split[i])
+            )
+
+        return tf.concat(y_split, axis=-1)
+
+    def group_stem(self, x):
+        x = tf.keras.layers.Conv2D(
+            32,
+            3,
+            strides=2,
+            padding="same",
+            use_bias=self.use_conv_bias,
+            kernel_initializer=self.kernel_initializer,
+        )(x)
+        x = self.norm(x)  # compare Fig 2 and Fig 3
+        x = self.act(x)
+
+        x = self.group_conv(x, 32, 3, 4)
+        x = self.norm(x)
+        x = self.act(x)
+
+        x = self.group_conv(x, 64, 3, 8)
+        x = self.norm(x)
+        x = self.act(x)
+
+        x = self.pool(x)
+
+        return x
+
+    def dense_block(self, x):
+        w = x
+        w = self.norm(w)
+        w = self.quant_conv(w, 64, 3)
+        y = tf.concat([x, w], axis=-1)
+        return y
+
+    def improvement_block(self, x):
+        w = x
+        w = self.norm(w)
+        w = self.quant_conv(w, 64, 3)
+        f_in = x.shape.as_list()[-1]
+        return tf.math.add(x, tf.pad(w, [[0, 0], [0, 0], [0, 0], [f_in - 64, 0]]))
+
+    def transition_block(self, x, filters):
+        x = self.norm(x)
+        x = self.pool(x)
+        x = self.act(x)
+        x = tf.keras.layers.Conv2D(
+            filters,
+            1,
+            use_bias=self.use_conv_bias,
+            kernel_initializer=self.kernel_initializer,
+        )(x)
+        return x
+
+    def block(self, x):
+        x = self.dense_block(x)
+        x = self.improvement_block(x)
+        return x
+
+    def build(self) -> tf.keras.models.Model:
+        x = self.input_tensor
+        x = self.group_stem(x)
+        for i, (n, f) in enumerate(zip(self.num_blocks, self.transition_features)):
+            for j in range(n):
+                x = self.block(x)
+            if f:
+                x = self.transition_block(x, f)
+
+        x = self.norm(x)
+        x = self.act(x)
+        x = tf.keras.layers.GlobalAvgPool2D()(x)
+        x = tf.keras.layers.Dense(
+            self.num_classes,
+            activation="softmax",
+            kernel_initializer=self.kernel_initializer,
+        )(x)
```

Comment:
This should make sure that sofmax is always computed in float32 when running the model in mixed precision mode. This might be the default in TF 2.0.0, but in the newer guide they recommend explicitly setting the dtype.
```suggestion
        x = tf.keras.layers.Dense(
            self.num_classes,
            kernel_initializer=self.kernel_initializer,
        )(x)
        x = tf.keras.layers.Activation("softmax", dtype="float32")(x)
```